virttest/qemu_devices/qbuses.py

```diff
@@ -650,6 +650,119 @@ def _increment_addr(self, addr, last_addr=None):
+
+    def is_direct_plug(self, device):
+        """
+        Justify if the device should be plug to pcie bus directly,
+        the device is not directly plugged only if it is a virtio
+        pci device and the device param pcie_direct_plug is "no".
+
+        :param device: the QBaseDevice object
+        :return: bool value for directly plug or not.
+        """
+        pcie_devices = ["virtio-net-pci", "virtio-blk-pci",
+                        "virtio-scsi-pci", "virtio-balloon-pci",
+                        "virtio-serial-pci", "virtio-rng-pci",
+                        "e1000e", "virtio-gpu-device"]
+        if device.get_param("driver") in pcie_devices and (
+                    device.get_param("pcie_direct_plug", "no") == "no"):
+            return False
+
+        return True
+
+    def add_root_port(self, addr, type, root_port=None, multifunction="off"):
+        """
+        Add pcie root port to the bus and __root_ports list, assign free slot,
+        chassis, and addr for it according to current list.
+
+        :param addr: addr to add the root port
+        :param type: root port type
+        :param root_port: pcie-root-port QDevice object
+        :param multifunction: if multifunction on or off
+        :return: pcie-root-port QDevice object
+        """
+        _addr = int(addr, 16)
+        if not root_port:
+            root_port = self._add_root_port(addr, type, multifunction)
+        bus = root_port.child_bus[0]
+        params = {'slot': _addr, 'chassis': _addr, 'addr': "0x%x" % _addr}
+        for key, value in params.iteritems():
+            root_port.set_param(key, value)
+        self.__root_ports[_addr] = bus
+        return root_port
+
+    def _add_root_port(self, addr, type, multifunction="off"):
+        """
+        Generate pcie-root-port QDevice object with certain addr and id
+
+        :param addr: addr to add the root port
+        :param type: root port type
+        :param multifunction: if multifunction on or off
+        :return: pcie-root-port QDevice object
+        """
+        _addr = int(addr, 16)
+        bus_id = "%s-root-port-%s" % (self.busid, _addr)
+        length = 1 if multifunction == "off" else 8
+        bus = QPCIBus(bus_id, 'PCIE', bus_id, length=length)
+        params = {'id': bus_id}
+        return qdevices.QDevice(type,
+                                params,
+                                aobject=bus_id,
+                                parent_bus=[{'aobject': self.aobject}],
+                                child_bus=bus)
+
+    def _insert(self, device, addr):
+        """
+        Inserts a pcie-root-port before the device if it is not a
+        direct plug device.
+
+        :param device: qdevice.QDevice object
+        :param addr: addr to insert the device
+        """
+        _addr = addr.split('-')[0]
+        if self.is_direct_plug(device):
+            driver = device.get_param("driver")
+            if driver in self.__root_type_list:
```

Comment:
Here I think it'd be better to always compare to all root port types as the fact that it's not supported won't make it non-root port. It'll always be a root port, only the qemu will fail to execute it, but it should be treated the same way as if supported (eg. for negative testing purposes)